larcorealg/Geometry/LineClosestPoint.tcc

```diff
@@ -0,0 +1,129 @@
+/**
+ * @file   larcorealg/Geometry/LineClosestPoint.tcc
+ * @brief  Utility for intersection of two 3D lines: template implementation.
+ * @author Gianluca Petrillo (petrillo@slac.stanford.edu)
+ * @see    larcorealg/Geometry/LineClosestPoint.h
+ */
+
+#ifndef LARCOREALG_GEOMETRY_LINECLOSESTPOINT_H
+# error "LineClosestPoint.tcc should not be included directly: #include \"larcorealg/Geometry/LineClosestPoint.h\" instead."
+#endif
+
+
+// LArSoft and framework libraries
+#include "cetlib/pow.h" // cet::square
+
+// C/C++ libraries
+#include "larcorealg/Geometry/geo_vectors_utils.h" // geo::vect::dot()
+
+// C++ standard library
+#include <cmath> // std::abs()
+#include <cassert>
+
+
+// -----------------------------------------------------------------------------
+template <typename Point, typename Vector>
+Point geo::LineClosestPoint(
+  Point const& startA, Vector const& dirA,
+  Point const& startB, Vector const& dirB,
+  std::pair<double, double>* locOnLines /* = nullptr */
+) {
+  /*
+   * The point on the first line ("A"):
+   *
+   *     p1(t) = c1 + t w1 (c1 the starting point, w1 its direction)
+   *
+   * has the minimal distance from the other line (c2 + u w2, with the same
+   * notation) at
+   *
+   *     t = [-(dc,w1)(w2,w2) + (dc,w2)(w1,w2)] / [ (w1,w2)^2 - (w1,w1)(w2,w2) ]
+   *     u = [ (dc,w2)(w2,w2) - (dc,w1)(w1,w2)] / [ (w1,w2)^2 - (w1,w1)(w2,w2) ]
+   *
+   * (where (a,b) is a scalar product and dc = (c1 - c2) ).
+   * If w1 and w2 are unit vectors, t and u are in fact the distance of the
+   * point from the "start" of the respective lines in "standard" geometry
+   * units.
+   * 
+   * If `locOnLines` is non-null, it is filled with { t, u }.
+   */
+
+  // aliases for quick notation
+  Point const& c1 = startA;
+  Vector const& w1 = dirA;
+
+  Point const& c2 = startB;
+  Vector const& w2 = dirB;
+
+  auto const dc = c2 - c1;
+
+  using geo::vect::dot;
+  double const dcw1 = dot(dc, w1);
+  double const dcw2 = dot(dc, w2);
+  double const w1w2 = dot(w1, w2); // this is cos(angle), angle between lines
+  assert(std::abs(std::abs(w1w2) - 1.0) >= 1e-10); // prerequisite: not parallel
+
+  using geo::vect::mag2;
+  double const w1w1 = mag2(w1);
+  double const w2w2 = mag2(w2);
+  double const inv_den = 1.0 / (cet::square(w1w2) - (w1w1 * w2w2));
+  double const t = ((dcw2 * w1w2) - (dcw1 * w2w2)) * inv_den;
+
+  if (locOnLines) {
```

Comment:
This is a bit too subtle.  Based on the interface, the user might expect the following call pattern:

```c++
std::pair<double, double>* locOnLines = nullptr;
auto pt = LineClosestPoint(startA, dirA, startB, dirB, locOnLines);
auto [t, u] = *locOnLines; // Boom!
delete locOnLines;
```

I think you're better off creating an overload set instead of providing a default value for the last argument:

```c++
auto pt1 = LineClosestPoint(startA, dirA, startB, dirB); // ........ // 4-arg. version
std::pair<double, double> locOnLines;
auto pt2 = LineClosestPoint(startA, dirA, startB, dirB, locOnLines); // 5-arg. version
assert(pt1 == pt2);
```

Reply:
Ok.
Although nobody should be expecting from LArSoft a C-flavoured interface like the one you describe.